pkg/cloudprovider/provider/vultr/provider.go

```diff
@@ -155,64 +245,188 @@ func (p *provider) Validate(ctx context.Context, _ *zap.SugaredLogger, spec clus
+	client := getClient(ctx, c.APIKey)
+	// Not looping on metadata assuming that tagged machines won;t cross
+	// pagination boundary
+	instances, _, resp, err := client.BareMetalServer.List(ctx, &govultr.ListOptions{
+		Tag: string(machine.UID),
+	})
 	if err != nil {
-		return nil, cloudprovidererrors.TerminalError{
-			Reason:  common.InvalidConfigurationMachineError,
-			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		return nil, vltErrorToTerminalError(resp.StatusCode, err)
+	}
+	resp.Body.Close()
+	for _, instance := range instances {
+		if sets.NewString(instance.Tags...).Has(string(machine.UID)) {
+			return &vultrPhysicalMachine{instance: &instance}, nil
 		}
 	}
+	return nil, cloudprovidererrors.ErrInstanceNotFound
+}
 
+func (p *provider) getVirtualMachine(ctx context.Context, c *Config, machine *clusterv1alpha1.Machine) (*vultrVirtualMachine, error) {
 	client := getClient(ctx, c.APIKey)
 
-	instances, _, err := client.Instance.List(ctx, &govultr.ListOptions{
+	instances, _, resp, err := client.Instance.List(ctx, &govultr.ListOptions{
 		Tag: string(machine.UID),
 	})
 	if err != nil {
-		return nil, vltErrorToTerminalError(err, "failed to list servers")
+		return nil, vltErrorToTerminalError(resp.StatusCode, err)
 	}
+	resp.Body.Close()
 
 	for _, instance := range instances {
-		for _, tag := range instance.Tags {
-			if tag == string(machine.UID) {
-				return &vultrInstance{instance: &instance}, nil
-			}
+		if sets.NewString(instance.Tags...).Has(string(machine.UID)) &&
+			instance.Label == machine.Name {
+			return &vultrVirtualMachine{instance: &instance}, nil
 		}
 	}
 
 	return nil, cloudprovidererrors.ErrInstanceNotFound
 }
 
 func (p *provider) Get(ctx context.Context, _ *zap.SugaredLogger, machine *clusterv1alpha1.Machine, _ *cloudprovidertypes.ProviderData) (instance.Instance, error) {
-	return p.get(ctx, machine)
+	c, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+	if !c.PhysicalMachine {
+		return p.getVirtualMachine(ctx, c, machine)
+	}
+
+	return p.getPhysicalMachine(ctx, c, machine)
 }
 
 func (p *provider) GetCloudConfig(_ clusterv1alpha1.MachineSpec) (config string, name string, err error) {
 	return "", "", nil
 }
 
+func (p *provider) waitForInstanceCreation(ctx context.Context, c *Config, instance instance.Instance, machine *clusterv1alpha1.Machine) error {
+	return wait.PollUntilContextTimeout(ctx, createCheckPeriod, createCheckTimeout, false, func(ctx context.Context) (bool, error) {
+		var err error
+		if !c.PhysicalMachine {
+			_, err = p.getVirtualMachine(ctx, c, machine)
+		} else {
+			_, err = p.getPhysicalMachine(ctx, c, machine)
+		}
+
+		if err != nil && err != cloudprovidererrors.ErrInstanceNotFound {
```

Comment:
When testing this latest push I saw that it skips this entire conditional, then returns true, but then the instance can't be found during reconciliation and it creates another.

I think the `err != cloudprovidererrors.ErrInstanceNotFound` prevents it from going into this retry loop correctly. It's basically the same check as `if cloudprovidererrors.IsNotFound(err)`. I removed the `err != cloudprovidererrors.ErrInstanceNotFound` in my local testbench and it looks to have created a single instance and waited correctly.

Reply:
Fixed.